### PR TITLE
[DataLoader] Add Example Only `fork` DataPipe

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60840 [WIP] Example of DataPipes and DataFrames integration
* #60680 [DataLoader] Add notebook with DataPipes API example
* **#60679 [DataLoader] Add Example Only `fork` DataPipe**
* #60678 [DataLoader] Decorate ZipDataPipe as `zip`
* #60675 [DataLoader] Add simple `groupby` DataPipe

This is example only DataPipe, not intended to be used in production. Used for tutorials, tests and documentation.
Have to be replaced by real `fork` upon DataLoader update.

Differential Revision: [D29461084](https://our.internmc.facebook.com/intern/diff/D29461084)

### CODE REVIEW

File: torch/utils/data/datapipes/iter/combining.py [hunk 1]
```diff
@@ -40,6 +40,26 @@ def __len__(self) -> int:
...
+        for i in self.buffer:
+            yield i
+
+
+@functional_datapipe('fork')
+class ForkIterDataPipe(IterDataPipe):
+
+    def __new__(cls, datapipe, instances):
+        result = []
+        buffer = [i for i in datapipe]
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/combining.py:59]: We should be able to use `list(datapipe)` now after fix the error from `__len__`.

**Review by ejguan — APPROVED** [review]
LGTM.

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461084).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461084).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461084).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461084).

**Pull request closed without merging**